docs/markdown/Python/python-goals/python-test-goal.md

```diff
@@ -196,6 +173,82 @@ python_tests(
+
+### Batching tests
+
+Running multiple `python_test`s within a single `pytest` process can sometimes improve test performance by allowing reuse of expensive high-level `pytest` fixtures. Pants allows users to opt into this behavior via the `batch_compatibility_tag` field on `python_test`, with the following rules:
+
+- If the field is not set, the `python_test` is assumed to be incompatible with all others and will run in a dedicated `pytest` process.
+- If the field is set and is different from the value on some other `python_test`, the tests are explicitly incompatible and are guaranteed to not run in the same `pytest` process.
+- If the field is set and is equal to the value on some other `python_test`, the tests are explicitly compatible and _may_ run in the same `pytest` process.
+
+Compatible tests _may not_ end up in the same `pytest` batch if:
+
+- There are "too many" tests with the same `batch_compatibility_tag`, as determined by the `[test].batch_size` setting.
+- Compatible tess have some incompatibility in Pants metadata (i.e. different `resolve` or `extra_env_vars`).
+
+Compatible tests that _do_ end up in the same batch will run in a single `pytest` invocation. By default the tests will run sequentially, but they can be parallelized by enabling `pytest-xdist` (see below). A single success/failure result will be reported for the entire batch, and additional output files (i.e. XML results and coverage) will encapsulate all of the included `python_test`s.
+
+> 📘 Tip: finding failed tests in large batches
+>
+> It can sometimes be difficult to locate test failures in the logging output of a large `pytest` batch. You can pass the `-r` flag to `pytest` to make this investigation easier:
+>
+> ```bash
+> ❯ ./pants test :: -- -r
+> ```
+>
+> This will cause `pytest` to print a "summary report" at the end of its output, including the names of all failed tests. See the `pytest` docs [here](https://docs.pytest.org/en/6.2.x/usage.html#detailed-summary-report) for more information.
+
+The high-level `pytest` fixtures that motivate batched testing are often defined in a `conftest.py` near the root of your repository, applying to every test in a directory tree. In these cases, you can mark all the tests in the directory tree as compatible using the [`__defaults__` builtin](doc:targets#field-default-values):
+
+```python BUILD
+python_test_utils()
+
+__defaults__({(python_test, python_tests): dict(batch_compatibility_tag="your-tag-here"),})
+```
+
+> 🚧 Caching batched tests
+>
+> Batched test results are cached together by Pants, meaning that if any file in the batch changes (or if a file is added to / removed from the batch) then the entire batch will be invalidated and need to re-run. Depending on the time it takes to execute your fixtures and the number of tests sharing those fixtures, you may see better performance overall by _not_ marking tests as batch-compatible, improving your cache-hit rate to skip running tests more often.
```

Comment:
Maybe mention tuning the batch size as an option? Not marking tests as batch-compatible (which is effectively batch size=1) is just one extreme of that.